src/levanter/eval_harness.py

```diff
@@ -0,0 +1,564 @@
+
+    @property
+    def EvalBatch(self):
+        return self.trainer.EvalBatch
+
+    @cached_property
+    def the_tokenizer(self):
+        return load_tokenizer(self.tokenizer)
+
+
+def run_lm_eval_harness(
+    config: LmEvalHarnessConfig,
+    model,
+    tokenizer,
+    EvalBatch,
+    axis_resources,
+) -> dict:
+    # tasks_to_run = tasks.get_task_dict(config.task_spec_or_default(), tasks.TaskManager())
+    tasks_to_run = config.to_task_dict()
+
+    outputs = _actually_run_eval_harness(config, model, tasks_to_run, tokenizer, EvalBatch, axis_resources)
+
+    return outputs
+
+
+def _actually_run_eval_harness(config: LmEvalHarnessConfig, model, tasks_to_run, tokenizer, EvalBatch, axis_resources):
+    max_examples = config.max_examples
+    max_eval_length = config.max_eval_length
+
+    EvalPos = model.Pos if max_eval_length is None else model.Pos.resize(max_eval_length)
+    harness = LevanterHarnessLM(EvalBatch, EvalPos, model, axis_resources, tokenizer)
+    # we always log_samples here and filter out the samples later if we don't want them
+    outputs = evaluator.evaluate(harness, tasks_to_run, limit=max_examples, log_samples=True)
+
+    averages = _compute_averages(outputs)
+    outputs["averages"] = averages
+
+    if not config.log_samples:
+        del outputs["samples"]
+
+    return outputs
+
+
+def _compute_averages(outputs):
+    """
+    Compute macro and micro averages of all metrics.
+
+    Args:
+        outputs: Dictionary with results and samples:
+                 - "results": Dictionary of task-level results.
+                 - "samples": Dictionary of task-level sample counts.
+
+    Returns:
+        Averages dictionary with macro and micro averages for all metrics.
+    """
+    averages = {}
+    metric_keys = set()
+
+    # Collect all possible metrics across tasks
+    for task_results in outputs["results"].values():
+        metric_keys.update(k for k in task_results.keys() if "stderr" not in k and k != "alias")
+
+    examples_per_task = [len(task_samples) for task_samples in outputs["samples"].values()]
+
+    # Compute macro and micro averages
+    for metric in metric_keys:
+        # Collect valid tasks for this metric
+        valid_tasks = [
+            (task_results.get(metric), examples_per_task[i])
+            for i, (task_name, task_results) in enumerate(outputs["results"].items())
+            if metric in task_results
+        ]
+
+        if not valid_tasks:
+            continue  # Skip metrics with no valid tasks
+
+        # Separate metric values and weights
+        metric_values, this_examples_per_task = zip(*valid_tasks)
+
+        # Compute macro and micro averages
+        averages["macro_avg_" + metric] = np.mean(metric_values)
+        averages["micro_avg_" + metric] = np.average(metric_values, weights=this_examples_per_task)
+
+    return averages
+
+
+NAT_TO_BIT = 1 / np.log(2)
+
+# eval_harness isn't consistent enough for this to actually be workable
```

Comment:
What does this mean?  lm evaluation harness is just a framework, so whether it's meaningful or not depends on the actual eval and the model size?

Reply:
well like, the samples for multiple choice tasks don't all have the answer in a standard format even though they easily could (and must internall at some point)